core/src/main/java/org/apache/accumulo/core/logging/ConditionalLogger.java

```diff
@@ -0,0 +1,193 @@
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiFunction;
+
+import org.apache.accumulo.core.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.Marker;
+import org.slf4j.event.Level;
+import org.slf4j.helpers.AbstractLogger;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+/**
+ * Logger that wraps another Logger and only emits a log message once per the supplied duration.
+ *
+ */
+public abstract class ConditionalLogger extends AbstractLogger {
+
+  private static final long serialVersionUID = 1L;
+
+  /**
+   * A Logger implementation that will log a message at the supplied elevated level if it has not
+   * been seen in the supplied duration. For repeat occurrences the message will be logged at the
+   * level used in code (which is likely a lower level). Note that the first log message will be
+   * logged at the elevated level because it has not been seen before.
+   */
+  public static class EscalatingLogger extends DeduplicatingLogger {
+
+    private static final long serialVersionUID = 1L;
+    private final Level elevatedLevel;
+
+    public EscalatingLogger(Logger log, Duration threshold, Level elevatedLevel) {
+      super(log, threshold);
+      this.elevatedLevel = elevatedLevel;
+    }
+
+    @Override
+    protected void handleNormalizedLoggingCall(Level level, Marker marker, String messagePattern,
+        Object[] arguments, Throwable throwable) {
+
+      if (arguments == null) {
+        arguments = new Object[0];
+      }
+      if (!condition.apply(messagePattern, Arrays.asList(arguments))) {
+        delegate.atLevel(level).addMarker(marker).setCause(throwable).log(messagePattern,
+            arguments);
+      } else {
+        delegate.atLevel(elevatedLevel).addMarker(marker).setCause(throwable).log(messagePattern,
+            arguments);
+      }
+
+    }
+
+  }
+
+  /**
+   * A Logger implementation that will suppress duplicate messages within the supplied duration.
+   */
+  public static class DeduplicatingLogger extends ConditionalLogger {
+
+    private static final long serialVersionUID = 1L;
+
+    public DeduplicatingLogger(Logger log, Duration threshold) {
+      super(log, new BiFunction<>() {
+
+        private final Cache<Pair<String,List<Object>>,Boolean> cache =
+            Caffeine.newBuilder().expireAfterWrite(threshold).maximumSize(250).build();
+
+        private final ConcurrentMap<Pair<String,List<Object>>,Boolean> cacheMap = cache.asMap();
+
+        /**
+         * Function that will return true if the message has not been seen in the supplied duration.
+         *
+         * @param msg log message
+         * @param args log message arguments
+         * @return true if message has not been seen in duration, else false.
+         */
+        @Override
+        public Boolean apply(String msg, List<Object> args) {
+          return cacheMap.putIfAbsent(new Pair<>(msg, args), true) == null;
```

Comment:
This LGTM now for adding to the cache, the only thing to point out is there is probably a chance the equals/hashcode could not be considered equal when it is if there are objects passed in the list for logging that are not implement hashcode/equals and the references are difference between logging runs. This is probably ok as it's best effort for the logging reduction but just wanted to point that out. If it's a problem in the future we could change it and just convert the objects to strings using toString() to compare.

Reply:
Yeah, I think best effort is fine. The penalty for when this check fails is more logging. But it's still likely less logging than now.